File: app/schemas/query.py
from __future__ import annotations

from datetime import datetime, timezone
from enum import Enum
from typing import Any

from .base import SchemaModel
from .connection import TableSchema
from .pydantic_compat import Field


class QueryStatus(str, Enum):
    SUCCESS = "success"
    FAILED = "failed"
    PARTIAL = "partial"


class ChartType(str, Enum):
    BAR = "bar"
    LINE = "line"
    PIE = "pie"
    TABLE = "table"


class QueryRequest(SchemaModel):
    question: str
    connection_id: str
    stream: bool = False
    page_number: int = 1
    page_size: int = 1000
    previous_query_id: str | None = None
    follow_up_instruction: str | None = None
    include_total_count: bool = False


class SQLExecutionRequest(SchemaModel):
    connection_id: str
    sql: str
    page_number: int = 1
    page_size: int = 1000
    include_total_count: bool = False


class PaginationInfo(SchemaModel):
    page_number: int = 1
    page_size: int = 1000
    offset: int = 0
    has_more: bool = False
    total_row_count: int | None = None
    applied_limit: int | None = None


class ExecutionResult(SchemaModel):
    columns: list[str] = Field(default_factory=list)
    rows: list[dict[str, Any]] = Field(default_factory=list)
    row_count: int = 0
    truncated: bool = False
    db_latency_ms: float | None = None
    total_row_count: int | None = None
    pagination: PaginationInfo | None = None
    source_sql: str | None = None


class ChartSuggestion(SchemaModel):
    chart_type: ChartType = ChartType.TABLE
    x_axis: str | None = None
    y_axis: str | None = None
    reason: str | None = None


class AgentStep(SchemaModel):
    step_type: str
    content: str
    timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    metadata: dict[str, Any] = Field(default_factory=dict)


class QueryHistoryItem(SchemaModel):
    id: str
    connection_id: str
    question: str
    rewritten_query: str | None = None
    retrieved_tables: list[str] = Field(default_factory=list)
    sql: str | None = None
    status: QueryStatus
    retry_count: int = 0
    llm_latency_ms: float | None = None
    db_latency_ms: float | None = None
    error_message: str | None = None
    error_type: str | None = None
    error_suggestion: str | None = None
    result_row_count: int | None = None
    context_source_query_id: str | None = None
    created_at: datetime


class QueryHistoryDetail(QueryHistoryItem):
    steps: list[AgentStep] = Field(default_factory=list)
    result_row_count: int | None = None
    result: ExecutionResult | None = None
    chart: ChartSuggestion | None = None
    retrieved_table_details: list[TableSchema] = Field(default_factory=list)
    telemetry: QueryTelemetry | None = None
    follow_up_context: str | None = None


class QueryTelemetry(SchemaModel):
    cache_hit: bool = False
    cache_key: str | None = None
    queue_wait_ms: float | None = None
    retrieval_latency_ms: float | None = None
    active_concurrency: int | None = None
    peak_concurrency: int | None = None
    schema_fingerprint: str | None = None
    schema_table_count: int | None = None
    llm_cache_hit: bool = False
    retrieval_backend: str | None = None
    embedding_backend: str | None = None
    retrieval_scope_count: int | None = None
    retrieval_candidates: int | None = None
    retrieval_selected: int | None = None
    retrieval_lexical_count: int | None = None
    retrieval_vector_count: int | None = None
    retrieval_top_score: float | None = None
    relationship_count: int | None = None
    relationship_tables: list[str] = Field(default_factory=list)
    column_annotation_count: int | None = None
    packed_context_tables: int | None = None
    packed_context_chars: int | None = None
    packed_context_truncated: bool = False
    packed_context_tokens: int | None = None
    packed_context_budget: dict[str, Any] = Field(default_factory=dict)
    packed_context_limit_reason: str | None = None
    packed_context_dropped_tables: list[str] = Field(default_factory=list)
    packed_context_dropped_columns: dict[str, list[str]] = Field(default_factory=dict)
    packed_context_dropped_relationship_clues: list[dict[str, Any]] = Field(default_factory=list)
    few_shot_example_ids: list[str] = Field(default_factory=list)
    tenant_isolation_key: str | None = None
    audit_id: str | None = None


class FollowUpContext(SchemaModel):
    history_id: str
    question: str
    connection_id: str
    rewritten_query: str | None = None
    retrieved_tables: list[str] = Field(default_factory=list)
    final_sql: str | None = None
    steps: list[AgentStep] = Field(default_factory=list)
    context_text: str


class QueryResult(SchemaModel):
    question: str
    rewritten_query: str
    retrieved_tables: list[TableSchema] = Field(default_factory=list)
    sql: str
    result: ExecutionResult | None = None
    chart: ChartSuggestion | None = None
    summary: str | None = None
    steps: list[AgentStep] = Field(default_factory=list)
    status: QueryStatus
    retry_count: int = 0
    llm_latency_ms: float | None = None
    db_latency_ms: float | None = None
    error_message: str | None = None
    error_type: str | None = None
    error_suggestion: str | None = None
    context_source_query_id: str | None = None
    telemetry: QueryTelemetry = Field(default_factory=QueryTelemetry)


class AnalyticsSummary(SchemaModel):
    recent_success_rate: float = 0.0
    average_llm_latency_ms: float = 0.0
    average_db_latency_ms: float = 0.0
    top_tables: list[dict[str, Any]] = Field(default_factory=list)
    error_distribution: list[dict[str, Any]] = Field(default_factory=list)


class TopTableItem(SchemaModel):
    table: str
    count: int


class ErrorDistributionItem(SchemaModel):
    type: str
    count: int
    percentage: float


class AnalyticsReport(SchemaModel):
    summary: AnalyticsSummary
    errors: list[ErrorDistributionItem] = Field(default_factory=list)
    top_tables: list[TopTableItem] = Field(default_factory=list)

File: NL2SQL Agent Frontend Development/src/app/types/query.ts
import type { TableSchema } from './connection';

export interface QueryRequest {
  connection_id: string;
  question: string;
  context_history?: string[];
  stream?: boolean;
  // 分页参数
  page_number?: number;
  page_size?: number;
  // Follow-up 支持
  previous_query_id?: string;
  follow_up_instruction?: string;
  // 其他选项
  include_total_count?: boolean;
}

// SQL 执行请求
export interface SQLExecutionRequest {
  connection_id: string;
  sql: string;
  page_number?: number;
  page_size?: number;
  include_total_count?: boolean;
}

// 导出请求
export interface ExportRequest {
  query_id?: string;
  connection_id?: string;
  sql?: string;
  format?: 'csv' | 'json' | 'excel';
}

export interface ExecutionResult {
  columns: string[];
  rows: Record<string, any>[]; // 修正为对象数组
  row_count: number;
  truncated?: boolean;
  db_latency_ms?: number;
  total_row_count?: number;
  pagination?: PaginationInfo;
  source_sql?: string;
  // legacy alias
  execution_time_ms?: number;
}

export interface ChartSuggestion {
  chart_type: 'bar' | 'line' | 'pie' | 'table';
  x_axis?: string;
  y_axis?: string;
  reason?: string;
}

export type StepType = 
  | 'rewrite'
  | 'schema_retrieval'
  | 'sql_generation'
  | 'sql_execution'
  | 'error_reflection'
  | 'retry'
  | 'done';

export interface PaginationInfo {
  page_number: number;
  page_size: number;
  offset?: number;
  has_more?: boolean;
  total_row_count?: number | null;
  applied_limit?: number | null;
  total_count?: number | null;
}

export interface AgentStep {
  step_type: StepType;
  content: string;
  timestamp: string;
  metadata?: Record<string, any>;
  status?: 'loading' | 'success' | 'error';
  error_type?: string;
  error_suggestion?: string;
  retry_count?: number;
  max_retries?: number;
}

export interface QueryResult {
  question: string;
  rewritten_query: string;
  retrieved_tables: TableSchema[];
  sql: string;
  result?: ExecutionResult | null;
  chart?: ChartSuggestion | null;
  summary?: string | null;
  steps: AgentStep[];
  status: 'success' | 'failed' | 'partial';
  retry_count: number;
  llm_latency_ms?: number | null;
  db_latency_ms?: number | null;
  error_message?: string;
  error_type?: string;
  error_suggestion?: string;
  context_source_query_id?: string;
  telemetry?: QueryTelemetry;
  // legacy aliases
  query_id?: string;
  sql_attempts?: number;
  results?: ExecutionResult | null;
  chart_suggestion?: ChartSuggestion | null;
  latency?: {
    llm_ms: number;
    db_ms: number;
  };
}

export type QueryStatus = QueryResult['status'];

export interface QueryTelemetry {
  cache_hit?: boolean;
  cache_key?: string | null;
  queue_wait_ms?: number | null;
  retrieval_latency_ms?: number | null;
  active_concurrency?: number | null;
  peak_concurrency?: number | null;
  schema_fingerprint?: string | null;
  schema_table_count?: number | null;
  llm_cache_hit?: boolean;
  retrieval_backend?: string | null;
  embedding_backend?: string | null;
  retrieval_scope_count?: number | null;
  retrieval_candidates?: number | null;
  retrieval_selected?: number | null;
  retrieval_lexical_count?: number | null;
  retrieval_vector_count?: number | null;
  retrieval_top_score?: number | null;
  relationship_count?: number | null;
  relationship_tables?: string[];
  column_annotation_count?: number | null;
  packed_context_tables?: number | null;
  packed_context_chars?: number | null;
  packed_context_tokens?: number | null;
  packed_context_truncated?: boolean;
  packed_context_budget?: Record<string, any>;
  packed_context_limit_reason?: string | null;
  packed_context_dropped_tables?: string[];
  packed_context_dropped_columns?: Record<string, string[]>;
  packed_context_dropped_relationship_clues?: Array<Record<string, any>>;
  few_shot_example_ids?: string[];
  audit_id?: string | null;
}

export interface Message {
  id: string;
  type: 'user' | 'assistant';
  content: string;
  sql?: string;
  generated_sql?: string;
  results?: ExecutionResult;
  execution_result?: ExecutionResult;
  chart_suggestion?: ChartSuggestion;
  chart?: ChartSuggestion;
  summary?: string;
  retrieved_tables?: TableSchema[];
  telemetry?: QueryTelemetry;
  steps?: AgentStep[];
  agent_steps?: AgentStep[];
  timestamp: Date;
  status?: 'loading' | 'success' | 'error';
  error_message?: string;
  error_type?: string;
  error_suggestion?: string;
  query_id?: string; // 关联的查询ID
  context_source_query_id?: string;
  retry_count?: number;
}

export interface QueryHistoryItem {
  id: string;
  connection_id: string;
  question: string;
  rewritten_query?: string;
  retrieved_tables?: string[];
  sql?: string;
  status: QueryResult['status'];
  retry_count?: number;
  error_message?: string;
  llm_latency_ms?: number;
  db_latency_ms?: number;
  created_at: string;
  error_type?: string;
  error_suggestion?: string;
  result_row_count?: number;
  context_source_query_id?: string;
  // legacy aliases
  user_question?: string;
  final_sql?: string;
  sql_attempts?: number;
  row_count?: number;
  execution_time_ms?: number;
  query_id?: string;
  natural_language_query?: string;
  generated_sql?: string;
}

// 查询上下文
export interface QueryContext {
  history_id: string;
  question: string;
  connection_id: string;
  rewritten_query?: string;
  retrieved_tables?: string[];
  final_sql?: string;
  steps?: AgentStep[];
  context_text: string;
  follow_up_context?: string;
  // legacy aliases
  query_id?: string;
  sql?: string;
  results_preview?: string;
}

export interface QueryHistoryDetail extends QueryHistoryItem {
  steps?: AgentStep[];
  result_row_count?: number | null;
  result?: ExecutionResult | null;
  chart?: ChartSuggestion | null;
  retrieved_table_details?: TableSchema[];
  telemetry?: QueryTelemetry;
  follow_up_context?: string | null;
}

export type QueryHistory = QueryHistoryDetail;
export type AssistantMessage = Message;
